### DISCUSSION TITLE
more precise controls for nested objects created via `-recurse`

### DISCUSSION THREAD

**Discussion by oodler577** [haukex/Util-H2O#14] [General]
This issue is more of a placeholder for thoughts on the potential for some precision controls over the nested objects created via `-recurse`. E.g., current _want_ is to specify default accessors in a nested hash at some arbitrary level.

```
my @default_accessors = (qw/
  level1_keyA->level2_keyA->level3_keyA
  level1_keyA->level2_keyA->level3_keyB
  level1_keyA->level2_keyA->level3_keyC
  level1_keyA->level2_keyB
  level1_keyA->level2_keyC
  level1_keyB
  level1_keyC
/);

$hashref = h2o -recurse, $hashref, @default_accessors;
```

A related, but additional "nice to have" is to be able control at depth the package. Upon thought, this extended syntax might get messy and I have no clean example; but here's a pass to DWIM:

```
my @ISA_map = (qw/
  level1_keyA,My::Foo::A
  level1_keyB,My::Foo::B
  level1_keyC,My::Foo::C
  level1_keyA->level2_keyA,My::Foo::A::A
  level1_keyA->level2_keyB,My::Foo::A::B
#... etc
/);

$hashref = h2o -recurse, -isa=>\@ISA_map, $hashref, @default_accessors;
```

The intent would not be to describe nested data structures like this extensively, but be able to provide a precise application for a few nested targets. This kind of reminds me of xpath or something crazy like that, and I definitely don't want to go in that direction. But I think some basic support would be super helpful. Any thoughts? I think a nice direction to do this would be in a "sparse" way, meaning design the description assuming a small number of these paths would need to be described rather than all of them (in the latter case, you may as well just initialize a DS explicitly).

Hope this makes sense...

**Comment by oodler577** [comment]
I'd also like to add, lack of this control forces me to do 2 things I don't want,

1. use the `-nolock` option of `h2o`
2. reference directly the hash key using deref syntax

The alternative would be to use `->can('KeyThatMayNotBeThere')` in place of `exists` and the hash key.

**Comment by haukex** [comment]
I've thought about the `-recurse` option in the past and back then came to the conclusion that trying to get more flexibility there would cause an exponential increase in the complexity of the options and in testing. Writing a quick `sub` to recursively apply an `h2o` call is not all too difficult (see for example my PerlMonks nodes [here](https://www.perlmonks.org/?node_id=1171313), [here](https://www.perlmonks.org/?node_id=1187920), [here](https://www.perlmonks.org/?node_id=11115752), and [here](https://www.perlmonks.org/?node_id=1183355)), and on Perl versions where it's available (5.16+, so 10 years now), [`__SUB__`](https://perldoc.perl.org/functions/__SUB__) is useful for writing recursive `sub`s. This might be something useful for the Cookbook.

However, this is not to say I'm not open to suggestions! I think the best thing would be if you could show an actual use case that would be an example for the kind of feature you're thinking of here, because at the moment I'm not feeling creative enough to come up with one myself.

**Comment by oodler577** [comment]
I think that the first step is going to be me implementing something in Util::H2O::More - I envisioned it as more of a proving ground and home for different things, and if they can help the core capability of `h2o` improve, then I'm super happy. This already happened with `baptise`. :-) ... I will also share with you a code privately that is better than just a creative example, it's a real code I've delivered. My approach basically now is to use `h2o` in as many real codes as I can to push the corners, my ultimate goal is to not have to drop into using deref syntax if at all possible. Thank you.

**Comment by oodler577** [comment]
I've created https://github.com/oodler577/p5-Util-H2O-More/issues/3 - I'll create wrapper around something probably inspired by your PM links above, but to also test out my line of thinking in the original post. Feel free to close this in lieu of https://github.com/oodler577/p5-Util-H2O-More/issues/3.

**Comment by haukex** [comment]
Sure, I'd be happy to take a look at your implementation in `::More` and some real-world examples to draw inspiration from! I'll leave this issue open for now as a reminder that at the very least I might want to put a recursion example in the cookbook.
